mage_ai/frontend/components/FileEditor/index.tsx

```diff
@@ -215,6 +230,52 @@ function FileEditor({
     </Spacing>
   );
 
+  const {
+    lastJsonMessage,
+    sendMessage,
+  } = useWebSocket(getWebSocket(), {
+    shouldReconnect: () => true,
+  });
+
+  useEffect(() => {
+    if (lastJsonMessage) {
+      // @ts-ignore
+      const jsonMessage: KernelOutputType = lastJsonMessage;
+      const executionState = jsonMessage?.execution_state;
+      const messageUUID = jsonMessage?.uuid;
+
+      if (messageUUID === DEFAULT_TERMINAL_UUID) {
+        if (ExecutionStateEnum.BUSY === executionState) {
+          setLoading(true);
+        } else if (ExecutionStateEnum.IDLE === executionState) {
+          setLoading(false);
+        }
+      }
+    }
+  }, [
+    lastJsonMessage,
+    setLoading,
+  ]);
+
+  const installPackagesButtonEl = (
+    <Spacing m={2}>
+      <KeyboardShortcutButton
+        inline
+        loading={loading}
+        onClick={() => {
+          openSidekickView(ViewKeyEnum.TERMINAL);
+          sendMessage(JSON.stringify({
+            code: `!pip install -r ~/../home/src/${projectName}/requirements.txt`,
```

Comment:
@wangxiaoyou1993 When user clicks the "Install packages" button, we run this command. Is this path reliable for getting to the project's `requirements.txt` file?

Reply:
it's not always this path. at backend, we get the path via
```
from mage_ai.data_preparation.repo_manager import get_repo_path
`get_repo_path()/requirements.txt`
```